File: Leetcode/919.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class CBTInserter {
private:
    vector<TreeNode*> v;
public:
    int getHeight(TreeNode* root) {
        if (root->left && root->right) {
            return 1+max(getHeight(root->left), getHeight(root->right));
        }
        else if (root->left && !root->right) {
            return 1+getHeight(root->left);
        }
        else if (!root->left && root->right) {
            return 1+getHeight(root->right);
        }
        return 1;
    }
    CBTInserter(TreeNode* root) {
        // bfs
        if (root) {
            // get height
            int height = getHeight(root);
            vector<vector<TreeNode*>> vec(height);
            queue<pair<TreeNode*,int> > q;
            q.push({root,0});
            while (!q.empty()) {
                auto node = q.front().first;
                int level = q.front().second;
                vec[level].emplace_back(node);
                q.pop();
                if (node->left) {
                    q.push({node->left,level+1});
                }
                if (node->right) {
                    q.push({node->right,level+1});
                }
            }
            v.emplace_back(nullptr);    // fake node in the first slot
            for (const auto &i : vec) {
                for (const auto &j : i) {
                    v.emplace_back(j);
                }
            }
        }
    }
    
    int insert(int val) {
        auto node = new TreeNode(val);
        v.emplace_back(node);
        int id = (int)v.size()-1;
        if (id & 1) {
            v[id / 2]->right = node;
        } else {
            v[id / 2]->left = node;
        }
        return v[id / 2]->val;
    }
    
    TreeNode* get_root() {
        return v[1];
    }
};

/**
 * Your CBTInserter object will be instantiated and called as such:
 * CBTInserter* obj = new CBTInserter(root);
 * int param_1 = obj->insert(val);
 * TreeNode* param_2 = obj->get_root();
 */

File: Leetcode/919.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
const MAXN int = 1005
type CBTInserter struct {
	v []*TreeNode
}

func getHeight(root *TreeNode) int {
	if root.Left != nil && root.Right != nil {
		a, b := getHeight(root.Left), getHeight(root.Right)
		if a > b {
			return 1+a
		}
		return 1+b
	} else if root.Left != nil && root.Right == nil {
		return 1+getHeight(root.Left)
	} else if root.Left == nil && root.Right != nil {
		return 1+getHeight(root.Right)
	}
	return 1
}

func Constructor(root *TreeNode) CBTInserter {
	res := CBTInserter{}
	res.v = nil
	if root != nil {
		vec := make([][]*TreeNode, getHeight(root))
		type Pair struct {
			first *TreeNode
			second int
		}
		queue := make(chan Pair, MAXN)
		queue <- Pair{root, 0}
		for len(queue) > 0 {
			pair := <- queue
			node := pair.first
			level := pair.second
			vec[level] = append(vec[level], node)
			if node.Left != nil {
				queue <- Pair{node.Left, level+1}
			}
			if node.Right != nil {
				queue <- Pair{node.Right, level+1}
			}
		}
		res.v = append(res.v, nil)
		for _, i := range vec {
			for _, j := range i {
				res.v = append(res.v, j)
			}
		}
	}
	return res
}

func (this *CBTInserter) Insert(val int) int {
	node := &TreeNode{Val: val, Left: nil, Right: nil}
	this.v = append(this.v, node)
	id := len(this.v)-1
	if (id & 1) != 0 {
		this.v[id >> 1].Right = node
	} else {
		this.v[id >> 1].Left = node
	}
	return this.v[id >> 1].Val
}

func (this *CBTInserter) Get_root() *TreeNode {
	return this.v[1]
}


/**
 * Your CBTInserter object will be instantiated and called as such:
 * obj := Constructor(root);
 * param_1 := obj.Insert(val);
 * param_2 := obj.Get_root();
 */